### PR TITLE
[1.19.x] Make the FireworkShapes enum extensible

### PR DESCRIPTION
**Pull request by danorris709**
This allows mods to extend the FireworkRocketItem.Shape enum so that they can render their own firework shapes whilst still using the firework logic (entity, and itemstack).

I changed the method in the Shape enum to use a for loop, rather than the map, as I believe the map is initialized too early for the enum to work otherwise. Not sure what other solution could be done, to reduce changes in the Minecraft code, but open to suggestions on that front.

[Example Video](https://i.gyazo.com/646d0369c89fda8f1221417e360f7261.mp4)

video

### CODE REVIEW

**Comment by ChampionAsh5357** [comment]
This implementation won't work in general as you cannot guarantee two extended enums will not have the same id. You would basically need to extend and replace all calls to store the string representing the firework instead. See how forge handles instances when storing the `MobEffect` when only the id is stored. We essentially patch in `forge:effect_id` to load the effect based on the registry name. It would be similar, but using a unique identifier representing the shape.

**Comment by danorris709** [comment]
Yeah I see what you mean about the IDs but I'm not sure what you mean in regards to looking at how the mob effect part works? That has the benefit of being a registry so it uses ResourceLocations as the ID whereas this only has a String name which would suffer the same issue as the integer IDs (people potentially having the same two IDs).

Unless you meant for me to change them to be ResourceLocations too?

**Comment by ChampionAsh5357** [comment]
A resource location is just a formatted string. So, we could either use the string name, assuming it doesn't have much of a purpose, or the enum name during extension and tell people the value must be unique. That should already be enforced for the enum name anyways, at least semantically, since added enum constants should be in the form '<modid>_<enum_name>'.

**Comment by danorris709** [comment]
> A resource location is just a formatted string. So, we could either use the string name, assuming it doesn't have much of a purpose, or the enum name during extension and tell people the value must be unique. That should already be enforced for the enum name anyways, at least semantically, since added enum constants should be in the form '_<enum_name>'.

Yes, apologies I didn't make it clear when I was comparing the ResourceLocation and String I understand that a ResourceLocation is a string but in best practice at least mods could use their mod id as a unique identifier. 

But yes I think you are correct that the enum name being the identifying string makes most sense here.

**Comment by danorris709** [comment]
I have updated it to what I think you were looking for, the test give command now looks like this:

`/give @p firework_rocket{Fireworks:{Explosions:[{"forge:firework_shape":pikachu,Colors:[I;14602026]}]}} 1`

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 1]
```diff
@@ -1595,4 +1596,20 @@ public static <T> HolderLookup.RegistryLookup<T> wrapRegistryLookup(final Holder
...
+
+    public static void saveFireworkShape(CompoundTag nbt, String key, FireworkRocketItem.Shape shape) {
+        nbt.putString(key, shape.name());
+    }
+
+    public static FireworkRocketItem.Shape loadFireworkShape(CompoundTag nbt, String key, @Nullable FireworkRocketItem.Shape fallback) {
+        String id = nbt.getString(key);
+
+        for (FireworkRocketItem.Shape value : FireworkRocketItem.Shape.values()) {
+            if (value.name().equals(id)) {
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java:1608]: Probably should check `equalsIgnoreCase` instead.

File: src/test/java/net/minecraftforge/debug/CustomFireworkShapeTest.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+
+import java.util.Objects;
+
+@Mod("custom_firework_shape_test")
+@Mod.EventBusSubscriber
+public class CustomFireworkShapeTest
```
Comment (ChampionAsh5357) [inline src/test/java/net/minecraftforge/debug/CustomFireworkShapeTest.java:20]: The entirety of this test class needs more documentation, especially with that double array dump.

File: src/main/java/net/minecraftforge/client/event/FireworkShapeRenderEvent.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+package net.minecraftforge.client.event;
+
+import net.minecraft.client.particle.Particle;
+import net.minecraft.world.item.FireworkRocketItem;
+import net.minecraft.world.phys.Vec3;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ *
+ * Fired when a non-default (aka non-minecraft) firework is attempting to be rendered
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/client/event/FireworkShapeRenderEvent.java:15]: This does not explain anything. What is being testing and what is the expected result of the test mod?

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Looks better, but I think that the event itself should be replaced with a registry that maps a shape to a functional interface with the method on the mod event bus. It could even be added to one of the existing registration events (e.g., `RegisterParticleProvidersEvent`).

Additionally, this does not do anything to allow crafting the shape onto a firework star. There should probably be an extension method on the item that returns a shape or null when it should default back to the initial list.

Also, do not use a copyrighted asset as a test mod.

**Comment by LexManos** [comment]
Pushed a solution, simple usecase is to create your particle shape in common code, then register your factory in client only code as it references client only classes.

**Pull request closed without merging**